### ISSUE TITLE
IWSLT 2017 Settings

### ISSUE THREAD

**Issue by stephanpeitz** [quanpn90/NMTGMinor#14]
Hi Ngoc-Quan,

I'm looking for more details about the settings used in your paper "Improving Zero-Shot Translation with Language-Independent Constraints", e.g. amount of GPUs, max. token per GPU, etc.

And what do you mean by "base learning rate 2"? "0.0002"?

Thanks a lot for your help!

Cheers,
Stephan

**Comment by quanpn90** [comment]
Hi Stephan,

Thank you for your interest towards my paper. I will try to give more details later since its quite late now. 
I trained all models in the paper with single GPUs. The pretrained models with 8 Transformer layers are done with 3584 tokens / minibatch for the 1080Ti and 4096 for the Titan XP with 1 more GB. Later on I was able to train faster with Titan Vs with 6000 tokens.

After pretraining, I kept training with the loss defined in the paper, but with 2048 tokens because the decoders  and encoders have to run twice. 

In the "attention is all you need" paper, the base learning rate is "1" and hidden in the learning rate formula. base learning rate simply decides the linear scale of the actual learning rate according to the formula (Noam). 

Thanks, and hope it helps,
Quan

**Comment by quanpn90** [comment]
Sorry I forgot to mention that the paper's source code has been moved to the DBMajor branch. It should provide the implementation all of the models mentioned in the paper

**Comment by stephanpeitz** [comment]
Thanks for quick reply! Very helpful! Cheers!

**Comment by stephanpeitz** [comment]
One follow-up question: 
How do you compute BLEU? Recased and detokenized with multi-bleu-detok?

Cheers,
Stephan

**Comment by quanpn90** [comment]
My reported BLEU score was done using detokenization followed by mteval13
script from Wmt. It was equivalent with using sacredbleu, I believe.

**Comment by stephanpeitz** [comment]
Sorry, one more question :)
I understand that you are using word features rather than an artificial token to perform multilingual NMT (and zero-shot translaton). 
From your Transformer (Decoder) code, it looks like that you only use these word features on the decoder side. Do you encode both source and target language or target language only?

**Comment by quanpn90** [comment]
Hi,

Yes I only used the language features on the decoder side. It is the target
language only.

For the project I added a special token for example <de> at the beginning
of each target sentence before doing binarization (preprocess). It will be
passed into the feature embedding and concatenated to each word embedding.
Personally I found that the zero-shot performance is consistently better
than using just a special <bos> token.

The source language is not encoded as reported. However due to the trials I
still added the <en> (for example en>de) to the source data but it is not
used in the model (It was at some point and then I removed it because I
think the encoder doesn't have to know the language its in).

I'm terribly busy thats why the scripts are not uploaded, but if you need I
can upload the scripts that I used for the project (and the data that I
preprocessed).

**Comment by quanpn90** [comment]
Hi,

Yes I only used the language features on the decoder side. It is the target
language only.

For the project I added a special token for example <de> at the beginning
of each target sentence before doing binarization (preprocess). It will be
passed into the feature embedding and concatenated to each word embedding.
Personally I found that the zero-shot performance is consistently better
than using just a special <bos> token.

The source language is not encoded as reported. However due to the trials I
still added the <en> (for example en>de) to the source data but it is not
used in the model (It was at some point and then I removed it because I
think the encoder doesn't have to know the language its in).

I'm terribly busy thats why the scripts are not uploaded (and also there are so many conflicts with the main branch ..) but I hope that helps.
